### ISSUE TITLE
Nested comments

### ISSUE THREAD

**Issue by ckeller** [Deducteam/Dedukti#146]
Hi
It would be very helpful to allow nested comments, e.g. `(; (; foo ;) ;)` .
Cheers

**Comment by francoisthire** [comment]
I don't know how to do that with menhir. Do you know any parser with menhir that has this feature?

**Comment by ckeller** [comment]
Actually, I think that comments are often handled by the lexer instead of the parser (and it seems to be the case for dedukti). One has to add a counter in the "comment" rule.
I know what you will say: patch welcome.

**Comment by Gbury** [comment]
As a reference, ocaml has nested comments.

**Comment by fblanqui** [comment]
The next version of Dedukti will be based on 
https://github.com/Deducteam/lambdapi. So, unless it is easy to 
implement, we won't change Dedukti's code anymore but Lambdapi's code 
instead.

Moreover, most editors provide ways to (un)comment a bunch of lines. So, 
it is not clear whether such a feature is really necessary.

Le 25/01/2019 à 17:35, Guillaume Bury a écrit :

**Comment by ckeller** [comment]
It is not very difficult to implement.
However, it is *very* useful: it happens all the time to comment a bunch of code that contains commented code! And you do want to keep the double comments.

**Comment by rlepigre** [comment]
Well, the real problem is multi-line comments. As always, simpler is better, that's why there are only single-line comments in the new syntax in `lambdapi`. One solution would then be to do the same in the legacy syntax of Dedukti. And this would solve the smiley problem as well.

Even if you allow nested multi-line comments, there are still loads of problems. For instance, how do you syntax-color them in editors? That's hard to do, and I do not know any editor that does it correctly for OCaml for instance: you should be able to comment any valid portion of code. And `(* ... "some string with *) inside" ... *)` is a valid comment in OCaml.

**Comment by GuillaumeGen** [comment]
I made a PR for this feature in #147 .
However, I thought it was a feature to be able to comment a definition with `(; a: Type. (; ;)` so to activate this definition you just have 2 characters to add: `(; ;) a: Type. (; ;)`.
(In fact, this feature is useful only if we write Dk files manually, which happens quite rarely)

**Comment by ckeller** [comment]
Thank you!
I am not sure what you mean. I am interested in writing for instance: `(; foo1 (; foo2 ;) foo3 ;)`
Some Dedukti files are nonetheless written by humans.

**Comment by GuillaumeGen** [comment]
I understand what you are interested in (and it is exactly what is implemented in #147 ), however, I have for instance written the file:
```
Nat : Type.

0 : Nat.
S : Nat -> Nat.

elt : Type.
Vect : Nat -> Type.

Nil : Vect 0.
Cons : n : Nat -> elt -> Vect n -> Vect (S n).

def Tail : n : Nat -> Vect (S n) -> Vect n.
[n, x, l] Tail n (Cons _ x l) --> l.

def S_bis : Nat -> Nat.
[]  S_bis 0         --> S (S 0)
[]  S_bis (S 0)     --> S (S 0)
[n] S_bis (S (S n)) --> S (S (S_bis n)).
Cons_bis : n : Nat -> elt -> Vect n -> Vect (S_bis n).

def Tail_bis : n : Nat -> Vect (S_bis n) -> Vect n.

(; La règle linéaire, qui pose un problème de préservation du typage 
[n, x, l] Tail_bis n (Cons_bis _ x l) --> l.
(; ;)
```
And I thought it was a feature that such a file is valid, since this file typechecks, but if you add `;)` right after "typage", you activate the ill-typed rule and can check that it is indeed refused by Dk.

**Comment by GuillaumeGen** [comment]
Solved in #147, merged by @francoisthire
